internal/framework/controller/reconciler.go

```diff
@@ -52,19 +52,23 @@ func NewReconciler(cfg ReconcilerConfig) *Reconciler {
 	}
 }
 
-func (r *Reconciler) newObject(objectType client.Object) client.Object {
+func (r *Reconciler) newObject(objectType client.Object) (client.Object, error) {
 	if r.cfg.OnlyMetadata {
 		partialObj := &metav1.PartialObjectMetadata{}
 		partialObj.SetGroupVersionKind(objectType.GetObjectKind().GroupVersionKind())
 
-		return partialObj
+		return partialObj, nil
 	}
 
 	// without Elem(), t will be a pointer to the type. For example, *v1.Gateway, not v1.Gateway
 	t := reflect.TypeOf(objectType).Elem()
 
 	// We could've used objectType.DeepCopyObject() here, but it's a bit slower confirmed by benchmarks.
-	return reflect.New(t).Interface().(client.Object)
+	obj, ok := reflect.New(t).Interface().(client.Object)
```

Comment:
I only fixed this `forcetypeassert` error, wondering what you thing about this.

Reply:
I think we can simplify this and panic if `ok == false`, since that would be a programming error. I'd rename the function to `mustCreateNewObject` to indicate that the function panics on error.